internal/utils/stats.go

```diff
@@ -20,156 +49,194 @@ type blockProcessorStats struct {
+func (s *BlockProcessorStats) UpdateProcessingTime(t time.Duration) {
 	s.processingTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateValidationTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateValidationTime(t time.Duration) {
 	s.validationTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateSkipListConstructionTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateSigValidationTime(t time.Duration) {
+	s.sigValidationTime.Observe(t.Seconds())
+}
+
+func (s *BlockProcessorStats) UpdateSkipListConstructionTime(t time.Duration) {
 	s.skipListConstructionTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateMerkelTreeBuildTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateMerkelTreeBuildTime(t time.Duration) {
 	s.txMerkelTreeBuildTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateCommitTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateCommitTime(t time.Duration) {
 	s.commitTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateCommitEntriesConstructionTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateCommitEntriesConstructionTime(t time.Duration) {
 	s.commitEntriesConstructionTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateStateTrieUpdateTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateStateTrieUpdateTime(t time.Duration) {
 	s.stateTrieUpdateTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateBlockStoreCommitTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateBlockStoreCommitTime(t time.Duration) {
 	s.blockStoreCommitTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateProvenanceStoreCommitTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateProvenanceStoreCommitTime(t time.Duration) {
 	s.provenanceStoreCommitTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateWorldStateCommitTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateWorldStateCommitTime(t time.Duration) {
 	s.worldstateCommitTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) updateStateTrieCommitTime(t time.Duration) {
+func (s *BlockProcessorStats) UpdateStateTrieCommitTime(t time.Duration) {
 	s.stateTrieCommitTime.Observe(t.Seconds())
 }
 
-func (s *blockProcessorStats) incrementTransactionCount(flag types.Flag, tx_type string) {
-	s.transactionCount.WithLabelValues(flag.String(), tx_type).Inc()
+func (s *BlockProcessorStats) UpdateTransactionsPerBlock(size int) {
+	s.transactionPerBlock.Observe(float64(size))
 }
 
-func (s *blockProcessorStats) updateBlockSize(size int) {
-	s.blockSize.Observe(float64(size))
+func (s *BlockProcessorStats) UpdateBlockSizeBytes(size int64) {
+	s.blockSizeBytes.Observe(float64(size))
 }
+
+func (s *BlockProcessorStats) IncrementTransactionCount(flag types.Flag, tx_type string) {
+	s.transactionCount.WithLabelValues(flag.String(), tx_type).Inc()
+}
+
+var Stats = newBlockProcessorStats()
```

Comment:
This is not a good approach, since we often run multiple instances of components in the same process. See replication, for example. Let's not use global variables but only an instance per component, as Senthil started implementing. It is OK for the performance evaluation, but not for production.

Reply:
I generally agree.

Note that Senthil implementation also contained a global stats object. This object was later assigned to each component. But it was still the same global object.
I simply exposed it even further to support the sig validation metric.

I'm not sure what Prometheus does if you define the same metric name twice.
So it is best to only create each Prometheus metric once (globally) and use it everywhere.
Prometheus handles that well.

We could define each metric in the appropriate module, instead of in a general module.
This will be a better code design, but in terms of correctness, this would not make any difference.